test/run/idl.mo

```diff
@@ -1,34 +1,37 @@
 //MOC-ENV MOC_UNLOCK_PRIM=yesplease
+//MOC-FLAG -v
 import Prim "mo:⛔";
 
-func serUnit() : Blob = (prim "serialize" : () -> Blob) ();
-func deserUnit(x : Blob) : () = (prim "deserialize" : Blob -> ()) x;
+func serUnit() : Blob = to_candid ();
+func deserUnit(x : Blob) : ?() = from_candid x;
 
-func serNats(x: Nat, y: Nat, z: Nat) : Blob = (prim "serialize" : (Nat,Nat,Nat) -> Blob) (x,y,z);
-func deserNats(x: Blob) : (Nat, Nat, Nat) = (prim "deserialize" : Blob -> (Nat,Nat,Nat)) x;
+func serNats(x: Nat, y: Nat, z: Nat) : Blob = to_candid (x,y,z);
+func deserNats(x: Blob) : ?(Nat, Nat, Nat) = from_candid x;
 
-func serText(x: Text) : Blob = (prim "serialize" : Text -> Blob) x;
-func deserText(x: Blob) : Text = (prim "deserialize" : Blob -> Text) x;
-
-func serBool(x: Bool) : Blob = (prim "serialize" : Bool -> Blob) x;
-func deserBool(x: Blob) : Bool = (prim "deserialize" : Blob -> Bool) x;
+func serBool(x: Bool) : Blob = to_candid (x);
+func deserBool(x: Blob) : ?(Bool) = from_candid x;
 
+func serText(x: Text) : Blob = to_candid (x);
+func deserText(x: Blob) : ?(Text) = from_candid x;
 
 Prim.debugPrint(debug_show (serUnit ()));
 Prim.debugPrint(debug_show (serNats (1,2,3)));
 Prim.debugPrint(debug_show (serText "Hello World!"));
 Prim.debugPrint(debug_show (serBool true));
 Prim.debugPrint(debug_show (serBool false));
 
+// unit and triples
+
+assert (?() == deserUnit (serUnit ()));
+assert(?(1,2,3) == (deserNats (serNats (1,2,3)) : ?(Nat,Nat,Nat)));
```

Comment:
@crusso This line fails

```
+IDL error: too few arguments NNN
+Error: failed to run main module `_out/idl.wasm`
+
+Caused by:
+    0: failed to invoke command default
+    1: wasm trap: unreachable
+       wasm backtrace:
+         0: @deserialize<NNN>
+         1: deserNats
+         2: init
+         3: _start
```

And I suspect that the issue is that we are desugaring `to_candid` so that it always introduces a single-component message, in this case a 3-tuple holding three `Nat`s.

This logic is mis-matched:

```
 | S.ToCandidE es ->
    (* desugaring to_candid introduces explicit IR tuple: *)
    let arg = tupE (exps es) in
    begin match T.normalize arg.note.Note.typ with
    | T.Tup ts -> (primE (I.SerializePrim [T.Tup ts]) [arg]).it
    | _ -> assert false
    end
```

Since we are not doing the same thing when deserializing now.

Reply:
There's something extremely confusing about the games that DeserializePrim is playing with its argument "list".

It is written as if it could work in an n-ary way, but it cannot, obviously.  All of those code paths (for typing the IR, and for generating code) are missing!  I think you warned me about this, but I didn't appreciate it earlier.

So we are *always* using this tuple-encoding trick, it seems.  But there's something strange happening that's preventing the trick from working here.

Reply:
> I actually do not see another way to get this to work

There is, I think, see below.